Implement support for handling OPTIONS message from client.

tolbertam opened the pull request:
Hello!  On the drivers team at DataStax, we are very interested in Scassandra and are beginning to pursue using it in our testing as it makes it much easier to test certain scenarios (i.e. timeouts, delays, certain exceptional conditions) then using a live cassandra instance.  While playing around with Scassandra I noticed that occasionally my clients would lose connection to a Scassandra instance.  This happened to have been caused by a recent unreleased change to java-driver.

All DataStax drivers have added or are in the process of adding a 'heartbeat' feature that sends an 'OPTIONS' message on a predefined interval to keep connections opened that may be timed out by ISPs or intermediate networking equipment that times out idle connections.  If a 'SUPPORTED' message is not sent in response within the driver's configured read timeout, the connection is closed and must be re-established.

The java driver will be implementing this feature in 2.0.10 and 2.1.5 (via [JAVA-533](https://datastax-oss.atlassian.net/browse/JAVA-533)) with a default idle time of 30 seconds.

I went ahead and implemented support for handling Options messages in Scassandra.  I also added some helper methods in CqlProtocolHelper for serializing & deserializing [string list], [string map], and [string multimap], which I afterwards realized may duplicate some functionality from CqlMap, CqlSet, and CqlList, but thought it could still be useful to break this functionality out into helper methods.

tolbertam commented on build.sbt:
```diff
@@ -86,7 +86,7 @@ pomExtra := {
   "com.typesafe.akka" % "akka-testkit_2.10" % akkaVersion % "test",
   "io.spray" % "spray-testkit" % sprayVersion % "test",
-  "com.datastax.cassandra" % "cassandra-driver-core" % "2.0.9" % "test" exclude("com.google.guava", "guava"),
+  "com.datastax.cassandra" % "cassandra-driver-core" % "2.0.10-SNAPSHOT" % "test" exclude("com.google.guava", "guava"),
```
Since HeartbeatVerificationTest depends on functionality from 2.0.10 (which is yet to be released), it might be a good idea to defer merging this until 2.0.10 is released (for convenience of running tests without needing to install the snapshot for everyone locally).

tolbertam commented:
The Travis CI build is failing because 2.0.10-SNAPHOST can not be resolved.   Since HeartbeatVerificationTest depends on functionality from 2.0.10 (which is yet to be released), it might be a good idea to defer merging this until 2.0.10 is released, but I figured I would open this earlier to make you aware of this change to the drivers and also to solicit feedback early as well.

chbatey commented:
Hey Sorry, I hadn't spotted this PR. Checking it out now.

chbatey commented:
Thanks for the heads up of the new driver feature. There is already some duplicate code between the Cql* classes and the protocol helper. This is because I am slowly moving to using the cassandra jar for message serialisation rather than doing it my self, with embedded collections coming etc I wanted to limit my work :)

Happy to delay merging this or if you want to put the test in a separate PR can take that along with a driver upgrade to 2.10

tolbertam commented:
Sorry for the delay in getting this all resolved.  I'm currently working on switching to using cassandra for serialization of list/map/multimap, that will simplify some of the things I did here quite a bit.  In addition, 2.0.10 is now released, so I'll make that change as well.  Shouldn't take me too much time, expect an update later today!

chbatey commented:
Thanks!

tolbertam commented:
Rebased on master and updated the code to use CBUtil from cassandra to serialize [string list], [string map], [string multimap] using CBUtil from Cassandra.  I considered using CqlList and CqlMap for this, but since those are for the cql columns and not frame contents it didn't seem right (I can make it work though).  Also removed the read methods I had previously added for these types as they were not used and upgraded to java-driver 2.0.10.  If everything looks good let me know and i'll squash my commits to 1 to clean the history up a bit.  Figured I'd leave it as is for now so it is more obvious what I changed.

chbatey commented:
Sorry for the delay - checking this out now!

chbatey commented:
Happy to merge this once you have squashed the commits. Thanks @tolbertam !

tolbertam commented:
Thanks @chbatey, I squashed and rebased against master, should be good to go!
